include/pstl_offload/internal/usm_memory_replacement_common.h

```diff
@@ -0,0 +1,134 @@
+
+static_assert(sizeof(__block_header) <= __header_offset);
+
+#if __linux__
+
+inline constexpr bool __is_power_of_two(std::size_t __number) {
+    return __number != 0 && (__number & __number - 1) == 0;
+}
+
+inline std::size_t __get_memory_page_size() {
+    static std::size_t __memory_page_size = sysconf(_SC_PAGESIZE);
+    assert(__is_power_of_two(__memory_page_size));
+    return __memory_page_size;
+}
+
+inline void* __allocate_shared_for_device(sycl::device* __device, std::size_t __size, std::size_t __alignment) {
+    // Unsupported alignment - impossible to guarantee that the returned pointer and memory header
+    // would be on the same memory page if the alignment for more than a memory page is requested
+    if (__alignment >= __get_memory_page_size()) {
+        return nullptr;
+    }
+
+    std::size_t __base_offset = std::max(__alignment, __header_offset);
+
+    // Check overflow on addition of __base_offset and __size
+    if (std::numeric_limits<std::size_t>::max() - __base_offset < __size) {
+        return nullptr;
+    }
+
+    // Memory block allocated with sycl::aligned_alloc_shared should be aligned to at least HEADER_OFFSET * 2
+    // to guarantee that header and header + HEADER_OFFSET (user pointer) would be placed in one memory page
+    std::size_t __usm_alignment = __base_offset << 1;
+    // Required number of bytes to store memory header and preserve alignment on returned pointer
+    // usm_alignment bytes are reserved to store memory header
+    std::size_t __usm_size = __size + __base_offset;
+
+    sycl::context __context = __device->get_platform().ext_oneapi_get_default_context();
+    void* __ptr = sycl::aligned_alloc_shared(__usm_alignment, __usm_size, *__device, __context);
+
+    if (!__ptr)
+        return nullptr;
+
+    void* __res = static_cast<char*>(__ptr) + __base_offset;
+    __block_header* __header = static_cast<__block_header*>(__res) - 1;
+    *__header = __block_header{__uniq_type_const, __ptr, __device, __size};
+    return __res;
+}
+
+inline bool __same_memory_page(void* __ptr1, void* __ptr2) {
+    std::uintptr_t __page_size = __get_memory_page_size();
+    std::uintptr_t __page_mask = ~(__page_size - 1);
+    std::uintptr_t __ptr1_page_begin = std::uintptr_t(__ptr1) & __page_mask;
+    std::uintptr_t __ptr2_page_begin = std::uintptr_t(__ptr2) & __page_mask;
+    return __ptr1_page_begin == __ptr2_page_begin;
+}
+
+inline auto __get_original_realloc() {
+    using _realloc_type = void* (*)(void*, std::size_t);
+
+    static _realloc_type __orig_realloc = _realloc_type(dlsym(RTLD_NEXT, "realloc"));
+    return __orig_realloc;
+}
+
+static void* __internal_realloc(void* __user_ptr, std::size_t __new_size) {
+    if (!__user_ptr) {
+        return std::malloc(__new_size);
+    }
+
+    __block_header* __header = reinterpret_cast<__block_header*>(__user_ptr) - 1;
+
+    if (__same_memory_page(__user_ptr, __header) && __header->_M_uniq_const == __uniq_type_const) {
```

Comment:
Let's fix `realloc` tomorrow. Deserves a separate discussion. At least we have the same bug with not resetting the usm constant

Reply:
I have added constant resetting

Reply:
Rewritten as discussed